gym_xiangqi/xiangqi_game.py

```diff
@@ -0,0 +1,144 @@
+    }
+    enemyCoord = {              # mapping of piece name to enemy initial positions
+        "king": [(0, 4)],
+        "queen": [(0, 3), (0, 5)],
+        "bishop": [(0, 2), (0, 6)],
+        "knight": [(0, 1), (0, 7)],
+        "rook": [(0, 0), (0, 8)],
+        "cannon": [(2, 1), (2, 7)],
+        "pawn": [(3, 0), (3, 2), (3, 4), (3, 6), (3, 8)]
+    }
+
+    def __init__(self):
+        # Xiangqi components
+        self.board = [[None for _ in range(self.boardCols)] for _ in range(self.boardRows)]
+        self.agentColor = random.randint(0, 1)
+        self.enemyColor = 0 if self.agentColor == 1 else 1
+        self.initAllPieces()
+
+        # PyGame components
+        self.running = False
+        self.winWidth = 900
+        self.winHeight = 1000
+        self.dim = (self.winWidth, self.winHeight)
+        self.display_surf = None
+
+
+    def initAllPieces(self):
+        # initialize agent and enemy pieces and place them on the board
+        for piece, pieceClass in self.classMap.items():
+            for i in range(self.pieceCount[piece]):
+                r = self.agentCoord[piece][i][0]
+                c = self.agentCoord[piece][i][1]
+                self.board[r][c] = pieceClass(self.agentColor, row=r, col=c)
+
+                r = self.enemyCoord[piece][i][0]
+                c = self.enemyCoord[piece][i][1]
+                self.board[r][c] = pieceClass(self.enemyColor, row=r, col=c)
+
+    def onInit(self):
+        """
+        Initialize/start the game with PyGame
+        ex. pygame.init()
+            (anything related to PyGame module that needs to be initialized)
+        """
+        pygame.init()
+        self.display_surf = pygame.display.set_mode(self.dim, pygame.HWSURFACE | pygame.DOUBLEBUF)
+        return True
+
+    def onEvent(self, event):
+        """
+        This routine is triggered when some kind of user/game event is detected
+        ex. when user closes the PyGame window
+            (mostly any keyboard/mouse input)
+        """
+        if event.type == pygame.QUIT:
+            self.running = False
+
+    def onUpdate(self):
+        """
+        Relfect detected changes and update game states
+        """
+        pass
+
+    def render(self):
+        """
+        Render current game state into graphics
+        """
+        pass
+
+    def cleanup(self):
+        """
+        Free resources and exit the game and
+        """
+        pygame.quit()
+
+    def reset(self):
+        """
+        terminate current game and create a new game
+        """
+        pass
+
+    def run(self):
+        """
+        Run the game until terminating condition is achieved
+        """
+        if self.onInit():
+            self.running = True
+
+        # TODO: this is just a high-level overview
+        while self.running:
+            for event in pygame.event.get():
+                self.onEvent(event)
+            self.onUpdate()
+        self.cleanup()
+
+
+if __name__ == "__main__":
```

Comment:
Do you want to put this in a separate unittest file instead?

Reply:
Yes definitely, but I want to look more into it before I actually write some test since I am not super familiar with writing unit tests. Is it okay if I merge this as it is for now?